src/transform/src/normalize_lets.rs

```diff
@@ -209,16 +218,34 @@ mod support {
     ) -> Result<(), crate::TransformError> {
         if let MirRelationExpr::LetRec { ids, values, body } = expr {
             for (id, value) in ids.iter().zip(values.iter_mut()) {
+                // Note that there can be an interleaving between steps of an outer and an inner
+                // WMR, i.e., it is not the case that we perform all steps of an inner WMR before
+                // performing a step in an outer WMR. This is fine, because we are in a lattice, so
+                // eventually we'll reach the same fixpoint.
                 refresh_types_helper(value, types)?;
                 let typ = value.typ();
-                let prior = types.insert(*id, typ);
-                assert!(prior.is_none());
+                let prior = types.insert(*id, typ.clone());
+                // Assert that the constraints not weakened. If this assert fails, then
+                // `refresh_types` might diverge.
+                // (Note that `prior` can't come from a shadowed binding, because there is no
+                // shadowing.)
+                assert!(prior.iter().all(|prior| {
```

Comment:
That is quite the mouthful. I am hitting this assertion . will try to simplify the query now.

Reply:
Test case
```DROP TABLE IF EXISTS t1;
CREATE TABLE t1 (f1 INTEGER PRIMARY KEY , f2 INTEGER NOT NULL, f3 INTEGER);
INSERT INTO t1 VALUES (1,1,1);
EXPLAIN WITH MUTUALLY RECURSIVE c1 (f1 INTEGER, f2 INTEGER, f3 INTEGER) AS (
	SELECT * FROM t1
	UNION ALL SELECT f3 + 3 + 1 AS f1 , f1 + 7 + 1 AS f2 , f2 + f3 + 1 AS f3 FROM c1 WHERE f3 < 100
) SELECT * FROM c1;
```
Please add to SLT once fixed.

Reply:
@ggevay I keep hitting that assertion no matter how much I dumb down the generated WMR queries. Could it be that it triggers on all queries that have the `+` operator? Unfortunately by using increment I ensure that the queries I run are halting, so it needs to work for testing correctness.

Reply:
Thanks for this test! The problem is that we run the new fixpoint loop not from `Top`, but from what a previous run of `NormalizeLets` gives us, which means that our nullability analysis would have to infer at least the same non-nullability now that it did in that earlier run of `NormalizeLets`, which seems almost impossible to ensure. For example, what happens in this query is that the earlier `NormalizeLets` operated on this:
```
Union
  Get u1
  Project (#3..=#5)
    Map (((#2 + 3) + 1), ((#0 + 7) + 1), ((#1 + #2) + 1))
      Filter (#2 < 100)
        Get l0
```
where non-nullability of `#2` inside the Map comes from the Filter, but in a later run of `NormalizeLets` we have the Filter and Map in the opposite order (due to a `CanonicalizeMfp` I guess):
```
Union
  Get u1
  Project (#3..=#5)
    Filter (#2 < 100)
      Map (((#2 + 3) + 1), ((#0 + 7) + 1), ((#1 + #2) + 1))
        Get l0
```
where we would need to propagate non-null knowledge backwards to infer the same non-nullability.

We could fix this by always starting the fixpoint loop from scratch (from `Top`) instead of what we currently have in the Get types, but @aalexandrov pointed out an other issue with the pessimistic approach, so I might completely rewrite the PR to the optimistic approach instead (i.e., 2. from https://github.com/MaterializeInc/materialize/issues/18553).